src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -362,6 +364,27 @@ class ConcatExpression : public detail::VariadicExpression {
   }
 };
 
+// ENCODE_FOR_URI
+[[maybe_unused]] auto encodeForUriImpl =
+    [](std::optional<std::string> input) -> IdOrString {
+  if (!input.has_value()) {
+    return Id::makeUndefined();
+  } else {
+    std::string value = input.value();
+
+    if (value.starts_with("\"")) {
+      auto contentEnd = ad_utility::findLiteralEnd(value, "\"");
+      if (contentEnd != 0) {
+        std::string content = value.substr(1, contentEnd - 1);
+        return boost::urls::encode(content, boost::urls::unreserved_chars);
+      }
+    }
+    return boost::urls::encode(value, boost::urls::unreserved_chars);
+  }
```

Comment:
In my understanding, the trailing quotes should already be gone as soon as this function is entered, but I'll have to check myself for all  cases. 
I know that their is strange stuff happening with the local vocabulary (when literals are manually created by expressions). 
Did you try out, which cases don't work if you don't do the `starts_with` check?

Reply:
In fact:
What we basically need is to consistently use a 
`LiteralOrIri` type that knows about whether a `string` is an iri or a literal, 
content, the datatype or langtag, and maybe some other meta information.
Would you be interested in working on this, then we can set up a meeting to get you started (some of the logic is already there for the literal case).

Reply:
I tested this by running the query below and using a debugging break point in line 373. 
* `BIND (encode_for_uri("Los Angeles") as ?encoded1) .` -> the `input.value().c_str()[0]` value is `L`
* `BIND (encode_for_uri("Los Angeles"^^xsd:string) as ?encoded2)` ->  the `input.value().c_str()[0]` value is `"`
Thus in the current implementation we need to check here for the trailing quotes. These need to be removed (line 378) otherwise they are included in the resulting encoding which is in my understanding unexpected behavior.

```
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT * WHERE {
  BIND (encode_for_uri("Los Angeles") as ?encoded1) .
  BIND (encode_for_uri("Los Angeles"^^xsd:string) as ?encoded2) .
  BIND (encode_for_uri("\"Los\" Angeles") as ?encoded3) .
}
```

Reply:
> What we basically need is to consistently use a LiteralOrIri type that knows about whether a string is an iri or a literal, content, the datatype or langtag, and maybe some other meta information. Would you be interested in working on this, then we can set up a meeting to get you started (some of the logic is already there for the literal case).

I would be highly interested in implementing this! I´d be available for a meeting next week Monday, Thursday or Friday to get onboarded to the task :)

Reply:
For these scheduling tasks please send me an email
(kalmbach@cs.uni-freiburg.de) so that I don't have to post private Zoom links to GitHub:)
If you have time on Monday late afternoon (4 pm or something later) that would be a good and early fit, but Thursday or Friday also works.